Detector/DetCommon/src/DetUtils.cpp

```diff
@@ -11,10 +23,160 @@ DD4hep::XML::Component det::getNodeByStrAttr(const DD4hep::XML::Handle_t& mother
+    if (aPreStepPoint) {
+      global = aStep.GetPreStepPoint()->GetPosition();
+    } else {
+      global = 0.5*(aStep.GetPreStepPoint()->GetPosition()+aStep.GetPostStepPoint()->GetPosition());
+    }
+    G4ThreeVector local  = aStep.GetPreStepPoint()->GetTouchable()->
+      GetHistory()->GetTopTransform().TransformPoint(global);
+    DD4hep::Geometry::Position loc(local.x()*MM_2_CM, local.y()*MM_2_CM, local.z()*MM_2_CM);
+    DD4hep::Geometry::Position glob(global.x()*MM_2_CM, global.y()*MM_2_CM, global.z()*MM_2_CM);
+    DD4hep::Geometry::VolumeManager::VolumeID cID = aSeg.cellID(loc,glob,volID);
+    return cID;
+  }
+  return volID;
+}
+
+std::vector<uint64_t> neighbours(DD4hep::DDSegmentation::BitField64& aDecoder,
+    const std::vector<std::string>& aFieldNames,
+    const std::vector<std::pair<int,int>>& aFieldExtremes,
+    uint64_t aCellId) {
+  std::vector<uint64_t> neighbours;
+  aDecoder.setValue(aCellId);
+  for(uint itField=0; itField < aFieldNames.size(); itField++) {
+    const auto& field = aFieldNames[itField];
+    int id = aDecoder[field];
+    if(id > aFieldExtremes[itField].first) {
+      aDecoder[field] = id-1;
+      neighbours.emplace_back(aDecoder.getValue());
+    }
+    if(id < aFieldExtremes[itField].second) {
+      aDecoder[field] = id+1;
+      neighbours.emplace_back(aDecoder.getValue());
+    }
+    aDecoder[field] = id;
+  }
+  return std::move(neighbours);
+}
+
+std::vector<std::pair<int,int>> bitfieldExtremes(DD4hep::DDSegmentation::BitField64& aDecoder,
+    const std::vector<std::string>& aFieldNames) {
+  std::vector<std::pair<int,int>> extremes;
+  int width = 0;
+  for(const auto& field: aFieldNames) {
+    width = aDecoder[field].width();
+    if( aDecoder[field].isSigned() ){
+      extremes.emplace_back(std::make_pair( ( 1LL << ( width - 1 ) ) - ( 1LL << width ),
+          ( 1LL << ( width - 1 ) ) - 1));
+    } else {
+      extremes.emplace_back(std::make_pair( 0,
+         (0x0001 << width) - 1 ));
+    }
+  }
+  return std::move(extremes);
+}
+
+CLHEP::Hep3Vector envelopeDimensions(uint64_t aVolumeId) {
+  DD4hep::Geometry::VolumeManager volMgr = DD4hep::Geometry::LCDD::getInstance().volumeManager();
+  auto pvol = volMgr.lookupPlacement(aVolumeId);
+  auto solid = pvol.volume().solid();
+  // get the envelope of the shape
+  TGeoBBox* box = (dynamic_cast<TGeoBBox*>(solid.ptr()));
+  // get half-widths
+  return CLHEP::Hep3Vector(box->GetDX(), box->GetDY(), box->GetDZ());
+}
+
+std::array<uint, 2> numberOfCells(uint64_t aVolumeId, const DD4hep::DDSegmentation::CartesianGridXY& aSeg) {
+  // // get half-widths
+  auto halfSizes = envelopeDimensions(aVolumeId);
+  // get segmentation cell widths
+  double xCellSize = aSeg.gridSizeX();
+  double yCellSize = aSeg.gridSizeY();
+  // calculate number of cells, the middle cell is centred at 0 (no offset)
+  uint cellsX = ceil((halfSizes.x()-xCellSize/2.)/xCellSize)*2+1;
+  uint cellsY = ceil((halfSizes.y()-yCellSize/2.)/yCellSize)*2+1;
```

Comment:
Why the `ceil` and `-yCellSize/2.` and `+1`? Shouldn't just the `ceil` without the other two do the trick?